docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+
+### Constructors and Destructors
+
+Verona distinguishes memory allocation with object construction.
+There are no `constructors` per se, in Verona, only static methods that can allocate memory (using `new`), build an object and return it.
+
+Example:
+```ts
+class Capibara
+{
+  let name : String;
+  var weight : F32; // Kgs
+
+  // Create a baby capibara
+  newborn(name : String) : Capibara
+  {
+    var self : Capibara = new Capibara;
+    self.name = name;
+    self.weight = 3; // Tiny baby!
+    self; // Returns `self`
+  }
+
+  eat(self: Capibara, food : F32)
+  {
+    self.weight += food;
+  }
+
+  poop(self: Capibara)
+  {
+    self.weight -= 0.1; // Tiny poop!
+  }
+};
+```
+
+On the example above, the `newborn` method creates a `Capibara` with a specific name and a default weight.
+The `name` can't change, so we declare it with `let`, while the `weight` does change, so it's a `var`.
+
+There's nothing special about the method `newbord` in comparison with others, apart from the fact that it allocates, initialises and returns an object of the same type.
+
+Destructors are still under debate, but they won't have to deallocate memory like in C++, just execute finalization tasks like closing file handles, running some cleanup routines, checking final consistency, etc.
+
+## Primitive Types
+
+Verona treats numeric types specially in the compiler but their representaiton isn't special in the language.
+
+Verona's numeric types are singleton types (no fields) that have their machine representaiton based on their names.
+So, `I32` is a signed integer with 32-bit width (in LLVM's parlour, an `i32`) and so on.
+
+The full list of primitive numeric types are:
+ * Unsigned integers: `U8`, `U16`, `U32`, `U64` and `U128`
+ * Signed integers: `I8`, `I16`, `I32`, `I64` and `I128`
+ * Floating Point: `F32` and `F64`
+ * System types: `ULength` and `ILength` (machine-word sized integers)
+
+Note, integer literals have no defined type and their types (and representation) will be inferred by the compiler depending on the type of the variable being assigned to.
+
+## Type Unions
+
+Type unions means a variable can be either type on the list.
+
+For example:
+```ts
+foobar(a : (A | B | C))
+{
+  ...
+}
+
+main()
+{
+  let a : A;
+  let b : B;
+  let d : D;
+
+  // This is fine
+  foobar(a);
+  foobar(b);
+
+  // This is not
+  foobar(d); // ERROR
+}
+```
+
+The imlementation of `foobar` will have to cope with all three types, either via a `match` statement or taking advantage of some behaviour that is common to all three (say, an interface they all implement).
+
+The compiler is free, however, to optimise the calls by creating two versions of `foobar` and replace the calls to the generic `foobar` for their specific ones.
+If the semantics is defined at compile time, this can simplifies a `match` statement into their specific matches and even inline the call altogether.
```

Comment:
```suggestion
If the semantics are defined at compile time, this can simplifies a `match` statement into their specific matches and even inline the call altogether.
```